plugins/postcss-light-dark-function/src/index.ts

```diff
@@ -0,0 +1,173 @@
+import postcssProgressiveCustomProperties from '@csstools/postcss-progressive-custom-properties';
+import type { PluginCreator, Source } from 'postcss';
+import { DARK_PROP, LIGHT_PROP, OFF, ON } from './props';
+import { colorSchemes } from './color-schemes';
+import { hasFallback } from './has-fallback-decl';
+import { hasSupportsAtRuleAncestor } from './has-supports-at-rule-ancestor';
+import { transformLightDark } from './transform-light-dark';
+
+const COLOR_SCHEME_REGEX = /color-scheme/i;
+const LIGHT_DARK_FUNCTION_REGEX = /light-dark\(/i;
+
+const basePlugin: PluginCreator<pluginOptions> = (opts) => {
+	const options: pluginOptions = Object.assign(
+		// Default options
+		{
+			preserve: true,
+		},
+		// Provided options
+		opts,
+	);
+
+	return {
+		postcssPlugin: 'postcss-light-dark-function',
+		prepare() {
+			let didTransformValues = false;
+			let transformedValueSource: Source | undefined = undefined;
+
+			return {
+				Declaration(decl, { atRule, rule }) {
+					const parent = decl.parent;
+					if (!parent) {
+						return;
+					}
+
+					if (COLOR_SCHEME_REGEX.test(decl.prop)) {
+						const [light, dark] = colorSchemes(decl.value);
+
+						if (light && dark) {
+							decl.cloneBefore({ prop: LIGHT_PROP, value: ON });
+							decl.cloneBefore({ prop: DARK_PROP, value: OFF });
+
+							const parentClone = parent.clone();
+							parentClone.removeAll();
+
+							parentClone.append(decl.clone({ prop: LIGHT_PROP, value: OFF }));
+							parentClone.append(decl.clone({ prop: DARK_PROP, value: ON }));
+
+							const prefers = atRule({ name: 'media', params: '(prefers-color-scheme: dark)', source: parent.source });
+							prefers.append(parentClone);
+
+							parent.after(prefers);
+
+							return;
+						}
+
+						if (dark) {
+							decl.cloneBefore({ prop: LIGHT_PROP, value: OFF });
+							decl.cloneBefore({ prop: DARK_PROP, value: ON });
+
+							return;
+						}
+
+						if (light) {
+							decl.cloneBefore({ prop: LIGHT_PROP, value: ON });
+							decl.cloneBefore({ prop: DARK_PROP, value: OFF });
+
+							return;
+						}
+
+						return;
+					}
+
+					if (LIGHT_DARK_FUNCTION_REGEX.test(decl.value)) {
+						if (hasFallback(decl)) {
+							return;
+						}
+
+						if (hasSupportsAtRuleAncestor(decl)) {
+							return;
+						}
+
+						const modified = transformLightDark(decl.value);
+						if (modified === decl.value) {
+							return;
+						}
+
+						didTransformValues = true;
+						if (!transformedValueSource) {
+							transformedValueSource = decl.source;
+						}
+
+						decl.cloneBefore({ value: modified });
+
+						if (decl.variable && decl.parent) {
+							const variableInheritanceRule = rule({
+								selector: '& *',
+								source: decl.source,
+							});
+
+							variableInheritanceRule.append(decl.clone({ value: modified }));
+							decl.parent.append(variableInheritanceRule);
+						}
```

Comment:
With nesting the parent might be an atrule, so we can't just clone that and manipulate its selector.

But because of nesting we can just insert a `& *` rule and it will be fine.

First time we get to use nesting to provide fallbacks instead of providing fallbacks for nesting itself 🎉